### PR TITLE
feat(history): [BACK-1439] Add `scheduledSurfaceHistory` subquery to ApprovedItem

### PR DESCRIPTION
**Pull request by nina-py**
## Goal

We need to add a query that retrieves scheduled entries for a corpus item. 

- Added a subquery to the admin schema.

- Added a resolver that takes in result limit + scheduled surface guid, both optional, and retrieves an array of scheduled items in return.

- [x] Tests ~outstanding~ done
- [x] Review feedback implemented

## References

JIRA ticket:
* https://getpocket.atlassian.net/browse/BACK-1439

### CODE REVIEW

File: schema-admin.graphql [hunk 1]
```diff
@@ -119,13 +119,6 @@ type ScheduledSurface {
     prospectTypes: [ProspectType!]!
 }
 
-"""
-Temp type to pacify federation checks
-"""
-type ApprovedCuratedCorpusItem @key(fields: "url") {
-    url: Url!
-}
-
```
Comment (nina-py) [inline schema-admin.graphql:128]: Unrelated change - I think this is safe to remove now that Prospect API has been updated to use the shortened `ApprovedCorpusItem` entity instead.

File: schema-admin.graphql [hunk 2]
```diff
@@ -205,6 +198,10 @@ type ApprovedCorpusItem @key(fields: "url") {
     A single sign-on user identifier of the user who last updated this entity. Null on creation.
     """
     updatedBy: String
+    """
+    Subquery to get the ScheduledCorpusItems to display to a user for a given date
+    """
+    scheduledSurfaceHistory(scheduledSurfaceGuid: ID, limit: Int): [ScheduledCorpusItem!]!
```
Comment (nina-py) [inline schema-admin.graphql:204]: Note that both filters are optional to future-proof this query. Right now the business case is to fetch this data for a single scheduled surface id + just the one most recent result, but eventually we will need to show on the frontend the entire history for a curated corpus item on demand, and that should show scheduled entries for all surfaces - either all or 6? 12? most recent ones.

File: schema-admin.graphql [hunk 3]
```diff
@@ -358,7 +355,7 @@ type ScheduledCorpusItem {
     """
     The associated Approved Item.
     """
-    approvedItem: ApprovedCorpusItem!
+    approvedItem: ApprovedCorpusItem
```
Comment (nina-py) [inline schema-admin.graphql:358]: I have made this optional to reuse this type in the return value of `scheduledSurfaceHistory` subquery. Open to suggestions here! Originally I thought we should set up a very abridged custom type that would only show date + surface guid, but then (see above notes on future-proofing it) decided to return the whole thing. 

But not the approved item! We already have it in the main query and it doesn't make sense to pull it in again.

File: src/database/queries/ApprovedItem.ts [hunk 4]
```diff
@@ -85,6 +85,36 @@ export async function getApprovedItemByUrl(
...
+): Promise<ScheduledItem[]> {
+  const approvedItem = db.approvedItem.findUnique({
+    where: { externalId },
+  });
+
+  return await db.scheduledItem.findMany({
+    take: limit,
+    orderBy: { scheduledDate: 'desc' },
+    where: {
+      approvedItemId: approvedItem.id,
```
Comment (nina-py) [inline src/database/queries/ApprovedItem.ts:110]: This line here breaks everything with a TS error at the moment. It's too late in the day to figure out what I'm doing wrong here.

File: schema-admin.graphql (hunk 2 above)
Comment (Herraj) [reply schema-admin.graphql:204]: Wondering if the result returned from this should be a `nullable` value? i.e `[ScheduledCorpusItem!]`, there could be a scenario where it hasn't been scheduled yet for the first time so this shouldn't return any history

File: src/admin/resolvers/queries/ApprovedItem.ts [hunk 5]
```diff
@@ -76,3 +77,53 @@ export async function getApprovedItemByUrl(
...
+  // the entire list of all the occasions this approved item was scheduled
+  // on all surfaces, in descending order (most recently scheduled first).
+  //
+  // Limiting it to a single surface means we'll only retrieve data for
+  // one scheduled surface (useful for when you only need the one to filter out
+  // prospects on the Prospecting page).
+  //
+  // Limiting the number of results to one means only the most recent result
+  // will be returned.
+  const { scheduledSurfaceGuid, limit } = args;
```
Comment (kschelonka) [inline src/admin/resolvers/queries/ApprovedItem.ts:103]: It seems unlikely that there would be a ton of entries to fetch, but I would recommend putting a default limit in place.

File: src/admin/resolvers/queries/ApprovedItem.ts [hunk 6]
```diff
@@ -76,3 +77,53 @@ export async function getApprovedItemByUrl(
...
+
+  // Check if the scheduled surface supplied is valid
+  const surface = ScheduledSurfaces.find((surface) => {
+    return surface.guid === scheduledSurfaceGuid;
+  });
+  if (!surface) {
```
Comment (kschelonka) [inline src/admin/resolvers/queries/ApprovedItem.ts:109]: IMO you could skip this check and just rely on no results being returned for the `where` condition

File: src/database/queries/ApprovedItem.ts (hunk 4 above)
Comment (Herraj) [reply src/database/queries/ApprovedItem.ts:110]: We just had to await the first db call when fetching the `approvedItem` and then do a non-null assertion, you were almost there!

File: src/admin/resolvers/queries/ApprovedItem.ts [hunk 7]
```diff
@@ -76,3 +77,53 @@ export async function getApprovedItemByUrl(
...
+    throw new UserInputError(
+      `Could not find Scheduled Surface with id of "${scheduledSurfaceGuid}".`
+    );
+  }
+
+  // Check if the limit specified passes basic sanity checks
+  if (limit.isNaN() || limit < 1) {
```
Comment (kschelonka) [inline src/admin/resolvers/queries/ApprovedItem.ts:116]: wish we could have finer-grained argument validation on the graphql side easily... almost all numeric args should be > 0

File: schema-admin.graphql (hunk 3 above)
Comment (kschelonka) [reply schema-admin.graphql:358]: I don't recommend making this nullable just based off the parent context. Better to use a custom type, return it as expected (and put in reasonable query depth limits to prevent weird circular resolvers), and use dataloaders to cache.

File: src/database/queries/ApprovedItem.ts (hunk 4 above)
Comment (kschelonka) [reply src/database/queries/ApprovedItem.ts:110]: Can you not do joins with prisma?

File: schema-admin.graphql (hunk 3 above)
Comment (bassrock) [reply schema-admin.graphql:358]: Agreed @kschelonka we should keep this required given the context and put in query depth limits

File: schema-admin.graphql (hunk 2 above)
Comment (nina-py) [reply schema-admin.graphql]: Agree these would be needed if we left the original `ScheduledCorpusItem` as a return type in place, but after a discussion with @Herraj and also feedback from @bassrock further down the review comments we decided a custom type that doesn't contain approved items would serve us better here.

Comment (nina-py) [reply schema-admin.graphql]: @Herraj - adding here for posterity as we discussed this in a pairing earlier today - the non-nullable return is for the frontend to have an easier time with return types. Something is always guaranteed to be returned, even if it's just an empty array.

Comment (bassrock) [reply schema-admin.graphql]: @nina-py I think I was unclear. I meant to imply return ScheduledCorpusItem but we can limit the query depth so a user can't infinitely recurse.

File: src/admin/resolvers/queries/ApprovedItem.ts (hunk 6 above)
Comment (nina-py) [reply src/admin/resolvers/queries/ApprovedItem.ts:112]: I could, but I would lose the ability to return an input error to the client, which is something I'd like to do to make the frontend side of things easier during development.

File: src/admin/resolvers/queries/ApprovedItem.ts (hunk 7 above)
Comment (nina-py) [reply src/admin/resolvers/queries/ApprovedItem.ts:119]: Thank you @kschelonka, that jogged my memory as I thought there was something out there to help with this. There's `NonNegativeInt` in `graphql-scalars`, trying it out on this field: https://www.graphql-scalars.dev/docs/scalars/non-negative-int

File: schema-admin.graphql (hunk 2 above)
Comment (bassrock) [reply schema-admin.graphql]: also just seeing the return type comment for frontend. makes sense

File: src/admin/resolvers/index.ts [hunk 8]
```diff
@@ -1,6 +1,10 @@
-import { DateResolver } from 'graphql-scalars';
+import { DateResolver, NonNegativeIntResolver } from 'graphql-scalars';
```
Comment (kschelonka) [inline src/admin/resolvers/index.ts:1]: 🙌

**Review by kschelonka — APPROVED** [review]
